<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <script>
      function all(promiseList) {
        return new Promise((resolve, reject) => {
          const resArr = [];
          promiseList.forEach((promise) => {
            promise.then(
              (res) => {
                resArr.push(res);
                if (resArr.length === promiseList.length) {
                  resolve(resArr);
                }
              },
              (err) => {
                reject(err);
              }
            );
          });
        });
      }
      const p1 = new Promise((resolve, reject) => {
        setTimeout(() => {
          resolve(11111);
        }, 1000);
      });

      const p2 = new Promise((resolve, reject) => {
        setTimeout(() => {
        //   reject(22222);
        resolve(22222);

        }, 2000);
      });

      const p3 = new Promise((resolve, reject) => {
        setTimeout(() => {
          resolve(33333);
        }, 3000);
      });

      // 需求: 所有的Promise都变成fulfilled时, 再拿到结果
      // 意外: 在拿到所有结果之前, 有一个promise变成了rejected, 那么整个promise是rejected
      all([p2, p1, p3])
        .then((res) => {
          console.log(res);
        })
        .catch((err) => {
          console.log("err:", err);
        });
    </script>
  </body>
</html>
